### PR TITLE
Implementation of audit log

### PR DESCRIPTION
**Pull request by igorsol**
This is implementation of audit log.
It also contains unit tests in jstests/audit which can be eexcuted with:
  python buildscripts/resmoke.py --audit
To execute tests one need to have mongoimport utility in the current directory.
Two tests are known to fail: audit_create_database.js and audit_replset_reconfig.js

### CODE REVIEW

**Comment by christianrober** [comment]
Looks good to me.  Team Note: In the future we will probably want to create a separate file for the Log classes, pulling them out of audit.cpp, but this is good enough for now.

File: src/mongo/db/audit/audit.cpp [hunk 1]
```diff
@@ -0,0 +1,901 @@
...
+                // If pwrite performs a partial write, we don't want to
+                // muck about figuring out how much it did write (hard to
+                // get out of the File abstraction) and then carefully
+                // writing the rest.  Easier to calculate the position
+                // first, then repeatedly write to that position if we
+                // have to retry.
+                fileofs pos = _file->len();
+
+                int writeRet;
+                for (int retries = 10; retries > 0; --retries) {
```
Comment (dbpercona) [inline src/mongo/db/audit/audit.cpp]: I'm not sure if we need this code here.   I cannot find a test for it.

Comment (igorsol) [inline src/mongo/db/audit/audit.cpp]: This part of code was ported from tokumx-audit-enterprise repository (https://github.com/percona/tokumx-audit-enterprise). There it was authored on Sep 29, 2014 as MXE-20 task. I don't know if MXE-20 can be viewed somewhere. History is here: https://github.com/percona/tokumx-audit-enterprise/commits/master/audit.cpp
In my understanding this code's purpose is to resolve issues when logRotate() switches current log file simultaneously with this write.

**Comment by dbpercona** [comment]
JIRA Ticket PSMDB-2

**Pull request merged**